antenna/util.py

```diff
@@ -211,25 +206,51 @@ def _retry_inner(fun):
         @wraps(fun)
         def _retry_fun(*args, **kwargs):
             attempts = 0
+            wait_times = wait_time_generator()
+
             while True:
+                # Grab the next wait time and if there isn't one, then this is
+                # the last attempt
+                try:
+                    next_wait = next(wait_times)
+                except StopIteration:
+                    next_wait = None
+
```

Comment:
So, this function got complicated by the fact that it does different things based on whether the last attempt threw an exception or returned a "failed result" coupled with the fact that when iterating over a generator, you will never know if the next iteration will halt.

To deal with that, for each iteration, we pull the next wait time before doing the iteration and in this way we know whether this is the last one or not. That's why this isn't in a `for next_wait in wait_time_generator():` kind of loop.